### DISCUSSION TITLE
3장

### DISCUSSION THREAD

**Discussion by rawfishthelgh** [somefood/cs-study#61] [공부만이 살길]
# 데이터베이스에 얽힌 돈 이야기

## 데이터베이스의 초기비용

초기비용 : 최초에 지급하는 돈

운영비용 : 서비스를 이용하는 기간에 계속해서 지급하는 돈

운영비용은 초기비용에 비해 낮게 설정된다 → 간과되는 경우가 있다…! 라고도 해석된다

### DBMS의 초기비용이란?

→ 소프트웨어의 라이센스 사용료를 의미, 그러나 통상적 개념의 가격과 라이센스 사용료는 다르다

판매 단위가 특수함 물리적 단위(~개)가 아닌 논리적 단위. 대표적으로는 다음 2개

1. 프로세서 라이선스 : dbms 설치 및 동작시키는 하드웨어 cpu 성능으로 가격 결정. 서버의 프로세서 수에 비례한다!
2. 사용자 라이선스 : dbms 이용하는 사용자 수로 가격 설정

→ 둘다 시스템 규모가 커질수록 사용료 올라간다! 왜냐면 둘 다 규모에 비례할 수 밖에 없음

→ 현실적으로 사용자 라이선스 적용은 힘들고 프로세서 라이선스 사용(사용자 수 파악은 어렵다)

오픈소스 dbms는 초기비용이 0원이라 인기 있는 것.

### 초기비용을 늘리는 범인

에디션과 에디선이 보유한 옵션이 관여

에디션은 대부분 두 가지가 존재

1. 스탠다드 에디션 : 주로 중소규모
2. 엔터프라이즈 에디션 : 주로 대규모
    1. 주요기능(옵션) :
        1. 신뢰성(가용성) : 클러스터 구성, 리플리케이션
        2. 성능 : 일정 수량 이상 프로세서 기술 지원, 테이블 파티셔닝, 성능 리포트, 데이터 압축
        3. 보안 : 데이터 암호화, 감사로그 기록

## 데이터베이스의 운영비용

db에서 운영비용 → 기술지원 비용을 의미(버그, db충돌시 복구 등…)

기술 Q&A, 버그 수정 프로그램 배포, 버전 업데이트, 새로운 os나 하드웨어 대응, 전문가 통한 문제해결, 노하우나 버스 정보 같은 기술 데이터베이스로의 접근 권리

### 기술지원

기술 지원 수준은 제품 배포를 기준으로 계속 내려감

- 제품 배포 시 → 제품 보급 전략 위해 다양한 기술지원 제공
- 시간 지나면서 서비스 레벨 낮아지고 신규 버그 패치 지원 x → 기술지원 종료
- 기술지원 종료 타이밍을 EOSL(End Of Service Life)라고 한다 → 어떤 소프트웨어를 채택할지를 검토할 때 자주 등장하는 단어다! (ex.사용하고 있는 제품 EOSL이 가까워지면 서버를 바꿈, EOSL이 가까운 제품 선택x)

즉, 기술지원 기간은 “구매시기”가 아닌 ”배포시기” 기준으로 결정!

그러나, 나온지 얼마 안됐다고 능사가 아니다. 이 경우 신뢰성에 결함 있는 경우가 높다. 상위 기능이 있고 기술지원이 긴 신버전, 신뢰성과 많은 정보를 가진 구버전의 트레이드 오프를 고민하자. 선행사례를 찾는 것이 도움이 됨

## 초기비용과 운영비용의 조합

즉 초기비용은 라이선스료, 운영비용은 기술지원 비용으로 정의할 수 있다.

상용 시스템에서는 초기비용과 운영비용이 모두 발생하는 경우가 일반적이나(Oracle, SQL server…), 오픈소스를 이용하면 초기비용이 발생하지 않고 운영비용만 발생한다(Mysql…).

### 오픈소스의 장단점

**장점**

오픈소스를 이용하면 초기 자본이 많지 않아도 이용 가능하다. 따라서 시작과 포기가 쉬워 의사결정 비용이 낮다. 타 db 마이그레이션 쉬움. 즉 심리적으로 쉽게 시도할 수 있다.

**단점**

서비스 제공 벤더가 망하거나 서비스 정지하면 이용 불가는 위험, 이용 기간이 길어지면 전체 비용이 높아질 수 있고, 요금 인상 등 비용 변경 요소 큼. 즉 **장기적으로 쓰기 불안한 점**이 많다

## 초기비용 트릭

보통 전체비용이 같더라도, 초기비용이 낮은 쪽을 심리적으로 선호하는 경향 있음. db도 마찬가지고 초기비용이 저렴하다고 무조건 선택하지 말고 운영비용을 포함한 전체비용을 확실히 계산하자

**Comment by JoungMinJu** [comment]
__초기비용__ 은 서비스를 구매할 때 최초에 지급하는 돈, __운영비용__ 은 상품이나 서비스를 이용하는 기간에 계속 지급하는 돈이다.

## 1. 데이터베이스의 초기 비용
시스템 개발의 **초기비용**으로서 대표적인 것은 서버나 네트워크 기기와 같은 HW 구매비용이나 프로그램 개발을 위해 엔지니어나 프로그래머에게 지급하는 급여를 들 수 있다. 
**운영비용**은 시스템 가동 후 장애 대응이나 프로그램 수정과 같은 유지보수 비용이 포함된다.

이때, __DBMS의 초기비용__ 은 __SW 라이센스 요금__ 이다. 이는 SW 사용허가료를 의미한다. 

라이선스료는 논리적인 단위로 판매된다. 대표적인 판매 단위는 두 가지가 존재한다.
__1. 프로세서 라이선스__ = DBMS를 설치해서 동작하는 HW의 CPU 성능에 따라 가격이 달라지는 라이선스 체계
__2. 사용자 라이선스__ = DBMS를 이용하는 사용자 수에 따라 가격이 결정되는 라이선스 체계입니다.

이 라이선스는 둘 다 DBMS가 동작하는 시스템의 규모가 커질수록 라이선스료도 올라간다는 특징이 있다. 시스템 규모가 커질수록 DB 서버의 프로세서 수는 증가하고, 이용하는 사용자 수도 많아지게 되기 때문이다.

그런데 어느 정도 규모를 가진 상용 시스템에서는 사용자 수를 파악하기가 어려워서 **프로세서 라이센스**를 사용하는 경우가 대부분이다. 현실적으로 **사용자 라이선스**를 적용하기 쉬운 곳은 개발 환경이나 시험 환경 같이 폐쇄된 네트워크에 설치되어 있거나 사용자 수를 파악할 수 있는 소규모 환경에 한정된다. 

PostgreSQL이나 MySQL 같은 오픈소스 DBMS는 라이선스료 자체가 무료인 경우가 있다. 이것은 해당 제품이 사용자에게 폭넓은 권한을 부여하는 라이선스 체계를 채택하고 있기 때문이다.

## 2. 초기 비용을 늘리는 주요 원인은?

### (1) 에디션과 옵션
대부분의 DBMS 제품은 다음 두 가지 에디션을 사용한다.

1. 스탠다드 에디션
2. 엔터프라이즈 에디션

기본적으로 "스탠다드 에디션"은 중소규모 시스템용. "엔터프라이즈 에디션"은 대규모 시스템용으로 구분되고 있다.

### (2) 에디션별 차이점
엔터프라이즈 에디션은 가격이 조금 더 높다.
여기에만 탑재되는 주요 기능은 다음과 같다. (종류마다 다를 수도 있음)

1. 신뢰성(가용성)
	= 클러스터 구성
    = 리플리케이션
2. 성능
	= 일정 수량 이상의 프로세서 기술 지원
    = 테이블 파티셔닝
    = 성능 리포트 출력
    = 데이터 압축
3. 보안
	= 데이터 암호화
    = 감사로그 기록
   
이러한 편리한 기능이 공짜는 아니기 때문에, 어디까지 고기능의 자원이 필요할지는 예산과 조율하면서 판단해야 한다.

## 3. DB의 운영 비용

### (1) 운영비용의 필요성
상용 시스템에서는 초기 비용 못지않게 운영비용이 발생한다. 
운영비용은 기간이 정해진 비용으로 "월 단위 또는 연단위"로 비용 계산을 한다. 즉, 이용 기간이 길수록 비용이 증가한다는 것이다.

### (2) DB의 운영비용
DB에서의 운영비용은?? __기술 지원 비용__ 이다. 치명적인 문제가 발생했을 때 기술적인 QnA부터 긴급 수정 프로그램(패치) 배포까지 DB 개발자의 지원 없이는 문제를 해결하기 어렵다. 이런 기술 지원 서비스에는 일반적으로 다음 항목이 포함된다.

1. 기술 QnA
2. 버그 수정 위한 프로그램 배포
3. 업데이트 관리
4. 새로운 OS나 HW 대응
5. 전문 기술자나 컨설턴트 통한 문제 해결
6. 노하우나 버그 정보 같은 기술 DB로의 접근 관리

기술 지원이 없는 SW로 상용 시스템을 만들어 운용하는 것은 치명적이다. 

__기술 지원에는 아래와 같은 특징__이 존재한다.

1. 기술 지원 수준은 매년 내려간다
2. 구매 시기와 기술지원 기간은 관계가 없다. 
3. 새로운 제품에는 신뢰성에 결함이 이을 수도 있다.


## 4. 초기비용과 운영비용의 조합

다음 세 가지의 조합을 생각해낼 수 있다. (상용 시스템에서)

1. 초기 비용 O + 운영 비용 O -> 가장 일반적
2. 초기 비용 O + 운영 비용 X -> 기술지원이 없는 것.
3. 초기 비용 X + 운영 비용 O -> 라이선스료 지급 안하고 기술 지원 비용만 발생

세 번째 선택지의 경우에 해당하는 대표적인 예는 __오픈소스 SW__ 이용하는 것이다. 
이 경우, 라이선스료는 무료로 하고 기술지원료만을 유상으로 하거나, 기한이 정해진 SW 사용권과 기술지원을 받을 수 있는 권리를 묶어서 판매하는 서브스크립션 형식을 따르는 것도 있다.

### (+) 서브스크립션 모델(임대 모델) vs 구매 모델
__서브스크립션__ 은 처음에 일정 금액을 가지고 있지 않아도 이용할 수 있단 장점이 있다. 시작도 간단하고 그만두는 것도 간단하다!(의사결정 비용이 낮다)

__구매 모델__ 은 더 장기적인 계획을 세울 수 있음이 장점이다.

**Comment by somefood** [comment]
# 3장 정리

## 왜 우리는 시스템에 돈을 내느가

- 시스템을 만드는 이유는 무엇일까? 좋은 서비스를 고객들에게 제공하는 것, 최종적으론 돈을 벌기 위함

## 데이터베이스 초기 비용

시스템을 구성할 때 발생하는 비용은 크게 두 가지로 나눌 수 있음

- 초기비용: 최초에 지급하는 돈. ex) 하드웨어 구매비용, 엔지니어나 프로그래머의 급여 등
- 운영비용: 서비스를 이용하는 기간에 계속해서 지급하는 돈. ex) 시스템 가동 후 장애 대응비, 유지보수 비용 등

데이터베이스의 초기 비용의 경우에는 소프트웨어 라이선스 요금이 들며, 하드웨어 성능에 따라 책정하는 `프로세서 라이선스`와 DBMS를 이용하는 사용자 수에 따라 결정되는 `사용자 라이선스`가 존재한다. 경우를 잘 파악해서 적절한 라이선스를 도입해야 한다.

## 초기비용을 늘리는 범인

초기비용이 늘어나는 요소로 에디션과 옵션을 들 수 있다. 잘못된 에디션이나 옵션 선택으로 인해 사용하지도 않는 불필요한 기능이 탑재되어 비용만 더 나갈 수 있으니 잘 파악하고 활용해야 한다. 엔터프라이즈 에디션의 경우에 아래와 같은 기능들을 제공할 수 있다.

- 신뢰성(가용성)
  - 클러스터 구성
  - 리플리케이션

- 성능
  - 일정 수량 이상의 프로세서 기술지원(서포트)
  - 테이블 파티셔닝
  - 성능 리포트 출력
  - 데이터 압축

- 보안
  - 데이터 암호화
  - 감사로그 기록

## 데이터베이스의 운영 비용

운영비용은 기간이 정해진 비용으로 `월 단위 또는 연 단위 얼마`로 정기적으로 발생하는 비용을 의미함. 데이터베이스에서 운영비용이 필요한 이유는 `기술지원 비용` 때문임. 유료 데이터베이스를 사용하다보며 크래시가 날 수 있는데, 이는 데이터베이스 개발자의 지원없이 해결하긴 어렵기에 관리비를 내는 개념으로 보면 된다. 또한 기술지원을 지출함으로써 장애 발생 시 `책임의 분산` 같은 보험적인 역할을 한다.

기술지원 서포트가 종료하는 타이밍을 EOSL(End of Servie Life)라고 한다. 기술지원 종료 시 버그를 해결할 수 없기에 도입 시 확인 필요하다. 그렇다고 너무 최신 버전을 사용하는 것이 좋은 점은 아니다. 소프트웨어는 언제나 버그가 발생할 수 있고, 최신 소프트웨어는 레퍼런스도 많이 없을 수 있기에 버전 선택은 신중해야 한다.

## 초기비용과 운영비용의 조합

초기비용과 운영비용 모델은 논리적으로 3가지 조합을 생각할 수 있음

- 초기비용 있음 + 운영비용 있음: 상용 시스템용에서 가장 일반적인 모델 (Oracle, SQL Server 등)
- 초기비용 있음 + 운영비용 없음: 유지보수 계약을 맺지 않아 기술지원이 없음. 현실성이 많이 떨어지는 모델
- 초기비용 없음 + 운영비용 있음: 라이선스료를 지급하지 않고 기술지원 비용만 발생. OSS를 사용하는 케이스

## 초기비용의 트릭

아래의 경우 중 어떤 것이 더 이득으로 보이는가?

1. 단말기 가격이 600,000원, 매월 통신료는 25,000원
2. 단말기 가격이 0원, 매월 통신료는 50,000원

많은 경우 2번이 이득으로 생각한다. 하지만 2년 사용량을 보았을 시 동일하며 이후에는 2번의 운영비용이 더 많이 나가게 된다. 이렇게 2번 같이 당장의 이득을 장기적인 이득보다 우선하는 것은 `현재지향 편향(Present Bias)`이라 한다. 비용을 산정할 때 이런 점도 주의해서 운영하도록 하자.

**Comment by seokhwan-an** [comment]
## 💡목표

데이터베이스의 ‘비용’에 대해 알아보자

### 데이터베이스의 초기비용

데이터베이스의 **초기 비용**은 소프트웨어 라이선스 요금(사용허가료)이다. 이 라이선스 비용은 두가지로 분리되는데 이는 각각 `프로세스 라이선스` 와 `사용자 라이선스` 가 있다.

`프로세스 라이선스`는 DBMS를 설치해서 동작하는 하드웨어 CPU 성능에 따라 가격이 결정되는 체계이며 `사용자 라이선스`는 DBMS를 이용하는 사용자 수에 따라 가격이 결정되는 라이선스 체계이다.

### 데이터베이스의 운영비용

데이터베이스에서 운영비용은 `기술지원 비용` 입니다.

기술 지원 서비스의 포함요소 

- 기술 Q&A
- 버그 수정을 위한 프로그램 배포
- 최신 버전으로의 업데이트 권리
- 새로운 OS나 하드웨어에의 대응
- 전문 기술자나 컨설턴트를 통한 문제 해걸
- 노하우나 버그 정보 같은 기술 데이터베이스로의 접근 권리

### 초기비용과 운영비용의 조합

- 초기비용 있음 + 운영비용 있음
- 초기비용 있음 + 운영비용 없음
- 초기비용 없음 + 운영비용 있음 → 오픈 소스를 이용하는 방식, 의사결졍 비용이 낮다.
